global-hooks/deckhouse-config/startup_sync_test.go

```diff
@@ -0,0 +1,423 @@
+			annotationJSON := fmt.Sprintf(`{"%s":"true"}`, shouldMigrateAnnotation)
+			Expect(generatedCM.Field("metadata.annotations").String()).Should(MatchJSON(annotationJSON))
+			Expect(generatedCM.Field("data.global").String()).Should(ContainSubstring("param1: val1"))
+			Expect(generatedCM.Field("data.deckhouse").Exists()).Should(BeTrue())
+			Expect(generatedCM.Field("data.certManager").Exists()).Should(BeTrue())
+			Expect(generatedCM.Field("data.certManagerEnabled").Exists()).Should(BeTrue())
+
+			// Test deployment
+			deckhouseDeploy := f.KubernetesResource("Deployment", "d8-system", "deckhouse")
+			Expect(deckhouseDeploy.Exists()).Should(BeTrue())
+			Expect(deckhouseDeploy.Field("spec.template.spec.containers.0.env.0.value").String()).Should(Equal(d8config.GeneratedConfigMapName), "should update deploy/deckhouse to use generated ConfigMap")
+		})
+	})
+
+	Context("Phase 2. Migrate to ModuleConfig objects", func() {
+
+		Context("giving valid ConfigMap", func() {
+			// Setup conversion chain for 'deckhouse' module.
+			var _ = conversion.RegisterFunc("deckhouse", 1, 2, func(settings *conversion.ModuleSettings) error {
+				return nil
+			})
+			var _ = conversion.RegisterFunc("deckhouse", 2, 3, func(settings *conversion.ModuleSettings) error {
+				return nil
+			})
+
+			BeforeEach(func() {
+				// Emulate migrated Deployment/deckhouse.
+				_ = os.Setenv("ADDON_OPERATOR_CONFIG_MAP", d8config.GeneratedConfigMapName)
+				cm := d8ConfigMap(d8config.GeneratedConfigMapName, `
+global: |
+  paramStr: "val1"
+  paramNum: 100
+deckhouse: |
+  paramStr: "val1"
+certManager: |
+  paramStr: "val1"
+certManagerEnabled: "false"
+unknownModule: |
+  paramBool: true
+`, shouldMigrateAnnotation)
+				f.KubeStateSet(cm)
+
+				f.BindingContexts.Set(f.GenerateOnStartupContext())
+				f.RunHook()
+			})
+
+			It("Should create ModuleConfig objects", func() {
+				Expect(f).To(ExecuteSuccessfully())
+
+				// Ensure phase 2.
+				Expect(f.LogrusOutput).Should(gbytes.Say("Migrate Configmap to ModuleConfig"), "should run phase 2")
+
+				// Check generated ConfigMap.
+				generatedCM := f.KubernetesResource("ConfigMap", "d8-system", d8config.GeneratedConfigMapName)
+				Expect(generatedCM.Exists()).Should(BeTrue())
+
+				annotations := generatedCM.Field("metadata.annotations")
+				Expect(annotations.Map()).Should(HaveLen(0), "should delete annotation, got %+v", annotations.String())
+
+				cfg := f.KubernetesGlobalResource("ModuleConfig", "global")
+				Expect(cfg.Exists()).Should(BeTrue(), "should create ModuleConfig/global")
+				Expect(cfg.Field("spec.enabled").Exists()).Should(BeFalse())
+
+				cfg = f.KubernetesGlobalResource("ModuleConfig", "deckhouse")
+				Expect(cfg.Exists()).Should(BeTrue(), "should create ModuleConfig/deckhouse")
+				Expect(cfg.Field("spec.enabled").Exists()).Should(BeFalse())
+				Expect(cfg.Field("spec.version").Int()).Should(BeEquivalentTo(3), "should migrate to latest version")
```

Comment:
Why spec.version equal 3? Is it should be 1?

Reply:
There is a registered conversion 2 -> 3 at the beginning of this Context. It is not a test for 'deckhouse' module.